gen/field.go

```diff
@@ -94,23 +105,88 @@ func (f fieldGroupGenerator) DefineStruct(g Generator) error {
 			<end>
 		}`,
 		f,
-		TemplateFunc("tag", func(f *compile.FieldSpec) string {
-			// We want to add omitempty if the field is an optional struct or
-			// primitive to redure "null" noise. We won't add omitempty for
-			// optional collections because omitempty doesn't differentiate
-			// between nil and empty collections.
-
-			if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
-				return fmt.Sprintf("`json:\"%s,omitempty\"`", f.Name)
-			}
-
-			return fmt.Sprintf("`json:%q`", f.Name)
-			// TODO(abg): Take go.tag and js.name annotations into account
-		}),
+		TemplateFunc("tag", generateTags),
 		TemplateFunc("declFieldName", f.declFieldName),
 	)
 }
 
+// generateTags parses the annotation on the thrift field and creates the resulting go tag
+func generateTags(f *compile.FieldSpec) string {
+	// start with default json tag
+	tk := []string{jsonTagName} // list of tag keys to preserve order
```

Comment:
You don't need to go through all this trouble to preserve ordering. Tag ordering doesn't matter since it's basically a key-value pair. We should just make sure it's deterministic by sorting the map keys before generating the Go tag string.